autopilot/contracts.go

```diff
@@ -39,3 +77,244 @@ func (ap *Autopilot) defaultContracts() ([]worker.Contract, error) {
 	}
 	return contracts, nil
 }
+
+func (ap *Autopilot) contractLoop() {
+	for {
+		select {
+		// TODO: use build var for loop interval
+		// TODO: add wakeChan (to trigger loop)
+		// TODO: run loop immediately
+		// TODO: logging
+		case <-ap.stopChan:
+			return
+		case <-time.After(time.Hour):
+		}
+
+		// don't run the contract loop if we are not synced
+		if !ap.store.Synced() {
+			continue
+		}
+
+		// fetch all contracts
+		cs, err := ap.defaultContracts()
+		if err != nil {
+			logErr(err)
+			continue
+		}
+
+		// run checks on the contract set
+		// TODO: dedupe existing contracts (?)
+		// TODO: check for IP range violations (?)
+
+		// run checks on the contracts individually
+		// TODO: separate loop (?)
+		csMap := make(map[string]int)
+		for i, c := range cs {
+			csMap[c.ID.String()] = i
+			// TODO: is host in host DB
+			// TODO: is max revision
+			// TODO: is offline
+			// TODO: is gouging
+			// TODO: is low score
+			// TODO: is out of funds
+			// TODO: update some contract metadata entity to reflect these checks
+		}
+
+		// gather some basic info
+		config := ap.store.Config()
+		needed := config.Contracts.Hosts
+		renewW := config.Contracts.RenewWindow
+		period := ap.currentPeriod()
+		renterAddress, err := ap.bus.WalletAddress()
+		if err != nil {
+			logErr(err)
+			continue
+		}
+
+		// TODO: check remaining funds before renew | form
+
+		// renew all renewable contracts
+		toRenew, err := ap.renewableContracts(renewW)
+		if err != nil {
+			logErr(err)
+			continue
+		}
+		for _, renew := range toRenew {
+			// renew contract
+			renterKey := ap.deriveRenterKey(renew.HostKey)
+			renewed, err := ap.renewContract(renew, renterKey, renterAddress, period)
+			if err != nil {
+				// TODO: handle error properly, if the wallet ran out of outputs
+				// here there's no point in renewing more contracts until a
+				// block is mined, maybe we could/should wait for pending transactions?
+				logErr(err)
+				continue
+			}
+
+			// swap contract in contract set
+			cs[csMap[renew.ID.String()]] = worker.Contract{
+				HostKey:   renew.HostKey,
+				HostIP:    renew.HostIP,
+				ID:        renewed.ID(),
+				RenterKey: renterKey,
+			}
+
+			// persist the contract
+			err = ap.bus.AddContract(renewed)
+			if err != nil {
+				logErr(err)
+				continue
+			}
+		}
```

Comment:
There should probably be a point somewhere in between renewing and forming contracts where we give up on renewing a contract ahead of its expiry and form a new one to avoid sudden data loss.

Reply:
resolving for now but keeping this in mind in the next PR 👍